src/github/actions-provider.ts

```diff
@@ -0,0 +1,32 @@
+/**
+ * Manage the versions used for GitHub Actions used in steps
+ */
+export class GitHubActionsProvider {
+  private actions = new Map();
+
+  /**
+   * Define an override for a given action.
+   * Setting an override for a specific action version will only override the exact same action version.
+   * To override all usages of an action, do not set version constraint.
+   *
+   * @example
+   * // Force any use of `actions/checkout` to use a specific commit
+   * project.github.actions.set("actions/checkout", "actions/checkout@ac59398");
+   *
+   * // Only replace usage of `v3` with a specific commit
+   * project.github.actions.set("actions/checkout@v3", "actions/checkout@ac59398");
+   */
+  public set(action: string, override: string): void {
+    this.actions.set(action, override);
+  }
+
+  /**
+   * Resolve an action name to the version that should be used,
+   * taking into account any overrides.
+   */
+  public get(action?: string): string | undefined {
```

Comment:
What is the use-case for `action` being `undefined`?

Also, it does not make sense to omit `action` here, even if we want to allow `undefined` as a convenience to avoid `null`-checking on call sites... Worst case scenario this should be:

```suggestion
  public get(action: string | undefined): string | undefined {
```

Reply:
Yes mostly to avoid unnecessary null checking. I see your point in keeping the public API clean though.